subsys/bluetooth/host/audio/iso.c

```diff
@@ -1066,6 +1068,8 @@ void bt_iso_recv(struct bt_conn *conn, struct net_buf *buf, uint8_t flags)
 			chan->ops->recv(chan, buf);
 		}
 	}
+
+	net_buf_unref(buf);
 }
```

Comment:
I think there is a bug in this code as it assemblies the data received into `chan->rx,` but then passes `buf` to the receiver.
For the `BT_ISO_END `case the `buf ` is already unref'ed - as the "tail" is copied into the existing "buf" stored in `chan->rx`.
I think the correct solution would be to pass `chan->rx,` to `chan->ops->recv` and then call  `net_buf_unref` on `chan->rx,`. IMO that should work for all cases (As I read the code it uses return in the cases where the SDU is not fully received.)

Reply:
Nice catch! I think you are correct and I've updated the PR.

Reply:
@Thalley This is also wrong, because conn->rx can be NULL when it is fragmented.

The code paths here are not easy to follow due to the weird structure.
The thing that Casper reacted to was that buf was unref'ed, but the pointer has also been reassigned (to point to conn->rx). So while it was correct for BT_ISO_END I think it was wrong for BT_ISO_SINGLE.
And now it is right for BT_ISO_SINGLE, but wrong for BT_ISO_END.

bt_conn_recv has also this structure, but it is easier to follow there because the reassigning of the buf pointer is done outside of the switch.
I think it has to be done like this to properly reset conn->rx to NULL, but not 100% sure about this

Reply:
@joerchan Your are right to some extend, but I think the problem is for the BT_ISO_START/SINGLE case - I missed that in my first review.

case BT_ISO_START:
case BT_ISO_SINGLE:
    conn->rx = buf; is missing for the SINGLE case.

Or you need to 
...
		if (chan->ops->recv) {
			chan->ops->recv(chan, conn->rx ? conn->rx : buf);
		}
		net_buf_unref(conn->rx ? conn->rx : buf);
But yes, you should probably set conn->rx = NULL after unref.

I would prefer to always set conn->rx - to simplify the RX-complete handling a bit.

Reply:
@Casper-Bonde-Bose Yes, we agree :) That's why I said it was wrong before with ISO_SINGLE, and now it is wrong for ISO_END.

I think the best would be to extract the common part of ISO_START and ISO_SINGLE to a function and the split those switch cases so that we get an easier code flow, and move the assignment of buf out of switch. Because this is just to complicated IMO.

Reply:
@joerchan @Casper-Bonde-Bose I've updated the code a bit more now. 

Basically it will now always store the `net_buf buf` in `conn->rx` and always unref the `buf`. The `buf` will be unref'ed in the `BT_ISO_CONT` and `BT_ISO_END` cases now, and `conn->rx` will just be assigned to the `buf`. 

At the end of the receive function, we are now using `bt_conn_reset_rx_state(conn)` (which we probably should have from the beginning), which unrefs the `buf` if single or the `buf` from `BT_ISO_START` otherwise.

Reply:
@Thalley looks better this way.

Just an FYI there is still a problem when the fragmentation of the ISO frame would split it in the header itself (for example length START with length 1). This has been handled for ACL, but missing handling for ISO.
The use-case for why the peer would fragment like this is for example to prioritize WIFI in a combination chip, and is something we have seen before.

But please fix that in a separate PR :)

Reply:
> @Thalley looks better this way.
> 
> Just an FYI there is still a problem when the fragmentation of the ISO frame would split it in the header itself (for example START fragment with length 1). This has been handled for ACL, but missing handling for ISO.
> The use-case for why the peer would fragment like this is for example to prioritize WIFI in a combination chip, and is something we have seen before.

There is no normative text about this actually, HCI ISO data packets don't use the same fragmentation scheme as ACL, in fact I do expect that the controller will never be sending less the 4 bytes due to ISO-AL sitting on the controller side as it will need to reconstruct the SDU before it can send to the host.

Reply:
@joerchan @Vudentz If there's a potential issue, I think we should be able to handle it, regardless of how unlikely it is (assuming that it's still spec valid). 

Looking at Core 5.2, Vol 4, Part E, 5.4.5 HCI ISO Data Packets, it contains the following text: 
> The first 4 octets of every HCI ISO Data packet contain a 4-octet Header field.

Where the the START/SINGLE/CONT/END is part of those 4 octets. Unless I'm misunderstanding something, I don't think a valid controller can send a HCI ISO data packet where the header is split.

Reply:
Yep, I recall having some discussion about that but while this could be assumed as normative the ACL text have been altered since the text was considered informative, so perhaps it worth raising the question with BT SIG core group if our interpretation is correct.